oauth/scope.js

```diff
@@ -0,0 +1,514 @@
+ * that do not open potential DoS vectors.  A naive item-by-item
+ * comparison of two sets can easily be accidentally quadratic.
+ *
+ * The key insight required for achieving this is that, for given
+ * a scope string `v`, we can enumerate the set of all possible values
+ * that imply `v`, and the size of this set of linear in the length
+ * of `v`.  These values are called the "implicants" of `v` and
+ * by pre-calculting and storing them, we are able to perform many
+ * `Scope` operations using simple set lookups.
+ *
+ * To that end, `Scope` instances store their data internally as a
+ * pair of cross-referenced maps, `this._values` and `this._implicants`,
+ * both of which are keyed by scope string values.
+ *
+ * The `_values` map has as its keys the actual scope values contained
+ * in the set, and the corresponding items are the set of implicants
+ * for each scope value.  Conversely, the `_implicants` map has as its
+ * keys every possible implicant of a value in the set, and the corresponding
+ * items are the set of scope values that are implied by that implicant.
+ *
+ * Thus:
+ *
+ *  - if `v in this._values` then the value `v` is a member of this
+ *    set, and `this._values[v]` gives all possible values that might
+ *    imply it.
+ *
+ *  - if `v in this._implicants` then the value `v` implies one or more
+ *    value in this set, and `this._implicants[v]` gives all the values
+ *    in this set that are implied by it.
+ *
+ * This scheme allows us to implement efficient membership checks, set
+ * comparison, and set union with deduplication of redundant values.
+ *
+ * The tradeoff here is memory usage, as we have to store the full
+ * set of implicants for each scope value.  We expect that this will not
+ * be too problematic in practice, because the size of this set is linear
+ * in the length of the scope value. A more memory-friendly implementation
+ * could use a trie-like data structure to efficiently store all implicants,
+ * with prefix compression.
+ *
+ *
+ */
+class Scope {
+
+  constructor(values, ...parentScopes) {
+    this._values = {};
+    this._implicants = {};
+    values.forEach(v => {
+      let implicants;
+      // If we're building a new `Scope` based on some existing parents,
+      // then we can probably find the implicants of each value as an
+      // existing set in one of those parents, rather than having to
+      // parse the value strings again.
+      for (const parentScope of parentScopes) {
+        implicants = parentScope._values[v];
+        if (implicants) {
+          break;
+        }
+      }
+      if (! implicants) {
+        implicants = new Set();
+        for (const implicant of getImplicantValues(v)) {
+          implicants.add(implicant);
+        }
+      }
+      this._addValue(v, implicants);
+    })
+  }
+
+  _addValue(value, implicants) {
+    // If this value is an implicant of some already-added value(s),
+    // then those values are redundant and can be removed.
```

Comment:
Can you explain why you remove the redundant values here? Are you trying to save memory, time on lookup, etc?